File: 543.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func diameterOfBinaryTree(root *TreeNode) (res int) {
	var dfs func(*TreeNode) int
	dfs = func(node *TreeNode) int {
		if node == nil {
			return -1
		}

		left := dfs(node.Left) + 1
		right := dfs(node.Right) + 1

		res = int(math.Max(float64(res), float64(left+right)))

		return int(math.Max(float64(left), float64(right)))
	}

	dfs(root)

	return
}

File: 543.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function (root) {
  let diameter = 0;

  function dfs(root) {
    if (!root) {
      return -1;
    }
    console.log("root.val", root.val);
    const left = dfs(root.left) + 1;
    const right = dfs(root.right) + 1;
    const sum = left + right;
    if (sum > diameter) {
      diameter = sum;
    }
    return Math.max(left, right);
  }
  dfs(root);

  return diameter;
};
